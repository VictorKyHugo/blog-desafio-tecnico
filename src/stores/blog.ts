import { ref } from 'vue';
import { defineStore } from 'pinia';
import * as BlogService from '@/services/blog.service';

interface IPosts {
  id: number;
  userId: number;
  title: string;
  content: string;
}

interface IUsers {
  id: number;
  name: string;
  username: string;
  email: string;
  phone: string;
}

interface IPostComments {
  id: number;
  postId: number;
  name: string;
  email: string;
  body: string;
}

export const useBlogStore = defineStore('blog', () => {
  const posts = ref<IPosts[]>([]);
  const users = ref<IUsers[]>([]);
  const comments = ref<IPostComments[]>([]);
  const currentPaginatePage = ref<number>(1);

  const fetchBlogData = async () => {
    return BlogService.fetchBlogData()
      .then((data) => {
        posts.value = data.posts;
        users.value = data.users;
      })
      .catch((err) => {
        throw new Error(err);
      });
  };

  const fetchPostComments = async (postId: number | string) => {
    return BlogService.fetchPostComments(postId)
      .then((data) => {
        comments.value = data.comments;
      })
      .catch((err) => {
        throw new Error(err);
      });
  };

  const getPostsFromUserById = (userId: number, quantity?: number, excludePostId?: number) => {
    let userPosts = [];

    userPosts = posts.value.filter((post) => post.userId === userId);

    if (excludePostId) {
      userPosts = userPosts.filter((post) => post.id !== excludePostId);
    }

    if (quantity) {
      // subtract 1 to match index
      userPosts = userPosts.slice(0, 4);
    }

    return userPosts;
  };

  const getPostsToRender = (itemsPerPage: number, currentPage: number) => {
    const pageFirstItemIndex = itemsPerPage * currentPage - itemsPerPage;
    const pageLastItemIndex = itemsPerPage * currentPage;

    return posts.value.slice(pageFirstItemIndex, pageLastItemIndex);
  };

  const getUserNameById = (id: number): string => {
    const [{ name }] = users.value.filter((user) => user.id === id);

    return name;
  };

  return {
    fetchBlogData,
    fetchPostComments,
    getPostsToRender,
    getUserNameById,
    getPostsFromUserById,
    posts,
    users,
    comments,
    currentPaginatePage
  };
});
