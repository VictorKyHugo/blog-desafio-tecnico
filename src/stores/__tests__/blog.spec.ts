import { setActivePinia, createPinia } from 'pinia';
import { useBlogStore } from '../blog';
import { describe, beforeEach, it, expect, vi } from 'vitest';

import * as BlogService from '../../services/blog.service';

const postsMock = [
  {
    id: 1,
    userId: 1,
    title: 'Title',
    content: 'Content'
  },
  {
    id: 2,
    userId: 2,
    title: 'Title',
    content: 'Content'
  },
  {
    id: 3,
    userId: 1,
    title: 'Title',
    content: 'Content'
  },
  {
    id: 4,
    userId: 2,
    title: 'Title',
    content: 'Content'
  }
];

const usersMock = [
  {
    id: 1,
    name: 'Flávio Ferreira',
    username: 'Flaferreira',
    email: 'flaferreira@getMaxListeners.com'
  },
  {
    id: 2,
    name: 'Luiz Silva',
    username: 'Luizin',
    email: 'luluzin@getMaxListeners.com'
  }
];

describe('Blog Store', () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it("should receive UserId and return it's name", () => {
    const store = useBlogStore();
    const userId = 1;

    const spy = vi.spyOn(store, 'getUserNameById').mockImplementation((id) => {
      const [{ name }] = usersMock.filter((user) => user.id === id);

      return name;
    });
    store.getUserNameById(userId);

    expect(spy).toHaveReturnedWith('Flávio Ferreira');
  });

  it('should call fetchBlogData from blogService', () => {
    const store = useBlogStore();
    const fetchBlogDataSpy = vi.spyOn(BlogService, 'fetchBlogData');

    store.fetchBlogData();

    expect(fetchBlogDataSpy).toHaveBeenCalled();
  });

  it('should call fetchPostComments from blogService', () => {
    const store = useBlogStore();
    const fetchPostCommentsSpy = vi.spyOn(BlogService, 'fetchPostComments');
    const postId = 1;

    store.fetchPostComments(postId);

    expect(fetchPostCommentsSpy).toHaveBeenCalled();
  });

  it('should get a quantity of post to render, according to currentPage and return the posts', async () => {
    const store = useBlogStore();

    const spy = vi.spyOn(store, 'getPostsToRender');
    spy.mockClear();

    const itemsPerPage = 2;
    const currentPage = 2;

    store.posts = postsMock;
    store.getPostsToRender(itemsPerPage, currentPage);

    expect(spy).toMatchSnapshot();
  });

  it('should get all posts from user', async () => {
    const store = useBlogStore();

    const spy = vi.spyOn(store, 'getPostsFromUserById');
    spy.mockClear();

    const userId = 2;

    store.posts = postsMock;
    store.getPostsFromUserById(userId);

    expect(spy).toMatchSnapshot();
  });

  it('should get 1 post from user', async () => {
    const store = useBlogStore();

    const spy = vi.spyOn(store, 'getPostsFromUserById');

    const userId = 2;

    store.posts = postsMock;
    const quantity = 1;
    store.getPostsFromUserById(userId, quantity);

    expect(spy).toMatchSnapshot();
  });

  it("should get 1 post excluding the first one by it's id", async () => {
    const store = useBlogStore();

    const spy = vi.spyOn(store, 'getPostsFromUserById');

    const userId = 1;

    store.posts = postsMock;
    const quantity = 1;
    const postIdToExclude = 1;
    store.getPostsFromUserById(userId, quantity, postIdToExclude);

    expect(spy).toMatchSnapshot();
  });
});
