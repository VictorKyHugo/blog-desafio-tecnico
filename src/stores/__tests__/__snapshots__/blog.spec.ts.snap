// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blog Store > should get 1 post excluding the first one by it's id 1`] = `
[MockFunction spy] {
  "calls": [
    [
      1,
      1,
      1,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": [
        {
          "content": "Content",
          "id": 3,
          "title": "Title",
          "userId": 1,
        },
      ],
    },
  ],
}
`;

exports[`Blog Store > should get 1 post from user 1`] = `
[MockFunction spy] {
  "calls": [
    [
      2,
      1,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": [
        {
          "content": "Content",
          "id": 2,
          "title": "Title",
          "userId": 2,
        },
      ],
    },
  ],
}
`;

exports[`Blog Store > should get a quantity of post to render, according to currentPage and return the posts 1`] = `
[MockFunction spy] {
  "calls": [
    [
      2,
      2,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": [
        {
          "content": "Content",
          "id": 3,
          "title": "Title",
          "userId": 1,
        },
        {
          "content": "Content",
          "id": 4,
          "title": "Title",
          "userId": 2,
        },
      ],
    },
  ],
}
`;

exports[`Blog Store > should get all posts from user 1`] = `
[MockFunction spy] {
  "calls": [
    [
      2,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": [
        {
          "content": "Content",
          "id": 2,
          "title": "Title",
          "userId": 2,
        },
        {
          "content": "Content",
          "id": 4,
          "title": "Title",
          "userId": 2,
        },
      ],
    },
  ],
}
`;
